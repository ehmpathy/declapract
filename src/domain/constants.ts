export enum RequiredAction {
  NO_CHANGE = 'NO_CHANGE',
  FIX_AUTOMATIC = 'FIX_AUTOMATIC',
  FIX_MANUAL = 'FIX_MANUAL',
}
