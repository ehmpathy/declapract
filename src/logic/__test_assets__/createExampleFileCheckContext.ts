import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
