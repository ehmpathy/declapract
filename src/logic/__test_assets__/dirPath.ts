export const testAssetsDirectoryPath = __dirname;
