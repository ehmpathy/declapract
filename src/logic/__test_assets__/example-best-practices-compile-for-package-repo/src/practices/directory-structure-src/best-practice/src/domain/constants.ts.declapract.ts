export const check = { optional: true }; // some projects don't have constants, and thats fine; this shows what naming should look like though
