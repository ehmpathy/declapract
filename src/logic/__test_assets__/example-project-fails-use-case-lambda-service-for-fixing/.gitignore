# all contents of dir are ignored since tests provision and fix the contents in here
*
!.gitignore
!declapract.use.yml

