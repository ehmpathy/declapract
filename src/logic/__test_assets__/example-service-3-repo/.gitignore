.declapract
node_modules
package-lock.json
