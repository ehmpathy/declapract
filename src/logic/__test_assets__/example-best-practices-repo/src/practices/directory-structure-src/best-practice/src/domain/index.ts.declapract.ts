import { FileCheckType } from '../../../../../../../../../domain';

export const check = FileCheckType.EXISTS;
