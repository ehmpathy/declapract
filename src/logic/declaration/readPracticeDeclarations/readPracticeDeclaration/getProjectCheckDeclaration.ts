import { FileCheckPurpose, FileCheckDeclaration } from '../../../../domain';
import { ProjectCheckDeclaration } from '../../../../domain/objects/ProjectCheckDeclaration';
import { readFileAsync } from '../../../../utils/fileio/readFileAsync';
import { listFilesInDirectory } from '../../../../utils/filepaths/listFilesInDirectory';
import { UnexpectedCodePathError } from '../../../UnexpectedCodePathError';
import { getFileCheckDeclaration } from './getFileCheckDeclaration/getFileCheckDeclaration';

export const getProjectCheckDeclaration = async ({
  purpose,
  declaredProjectDirectory,
}: {
  purpose: FileCheckPurpose;
  declaredProjectDirectory: string;
}): Promise<ProjectCheckDeclaration> => {
  // grab name from the directory
  const name = (() => {
    if (purpose === FileCheckPurpose.BEST_PRACTICE)
      return (new RegExp(/\/([\w-]+)\/best-practice$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
    if (purpose === FileCheckPurpose.BAD_PRACTICE)
      return (new RegExp(/bad-practices\/([\w-]+)$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
  })();
  if (!name)
    throw new UnexpectedCodePathError(
      `neither best-practice name nor bad-practice name was extractable from the declared project directory '${declaredProjectDirectory}'`,
    );

  // grab paths to _all_ files in this dir (not just at root level)
  const paths = await listFilesInDirectory({ directory: declaredProjectDirectory });

  // grab the meta files (i.e., path matches `${projectRoot}/.declapract.*`)
  const metaFilePaths = paths.filter((path) => new RegExp(/^\.declapract\./).test(path));

  // group all of the other files by main file name (i.e., key = filePath.replace('.declapract.ts$', ''))
  const projectFilePaths = [
    ...new Set(
      paths.filter((path) => !metaFilePaths.includes(path)).map((path) => path.replace(/\.declapract\.ts$/, '')),
    ),
  ];

  // for each "main file", get the FileCheckDefinition, now that we have all the files defined for it
  const checksAndErrors = await Promise.all(
    projectFilePaths.map((declaredFileCorePath) =>
      getFileCheckDeclaration({ purpose, declaredProjectDirectory, declaredFileCorePath }).catch((error) => error),
    ),
  );
  const anError = checksAndErrors.find((checkOrError) => checkOrError instanceof Error);
  if (anError) throw anError;
  const checks: FileCheckDeclaration[] = checksAndErrors.filter((checkOrError) => !(checkOrError instanceof Error));
  if (!checks.length) {
    console.log({ anError });
    console.log({
      declaredProjectDirectory,
      projectFilePaths,
      checksAndErrorsLength: checksAndErrors.length,
      checksLength: checks.length,
    });
  }

  // get readme contents, if readme defined
  const readme = metaFilePaths.includes('.declapract.readme.md')
    ? await readFileAsync({ filePath: `${declaredProjectDirectory}/.declapract.readme.md` })
    : null;

  // return the project def
  return new ProjectCheckDeclaration({
    name,
    checks,
    readme,
  });
};
