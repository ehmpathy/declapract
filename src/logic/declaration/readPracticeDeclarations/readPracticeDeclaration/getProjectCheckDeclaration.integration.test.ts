import { FileCheckPurpose } from '../../../../domain';
import { testAssetsDirectoryPath } from '../../../__test_assets__/dirPath';
import { getProjectCheckDeclaration } from './getProjectCheckDeclaration';

describe('getProjectCheckDeclarationFromDirectory', () => {
  it('should define correctly when only one root level file check and a readme', async () => {
    const project = await getProjectCheckDeclaration({
      purpose: FileCheckPurpose.BEST_PRACTICE,
      declaredProjectDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times/best-practice`,
    });
    expect(project.checks.length).toEqual(1);
    expect(project).toMatchSnapshot();
  });
  it('should define correctly when many files in many levels', async () => {
    const project = await getProjectCheckDeclaration({
      purpose: FileCheckPurpose.BEST_PRACTICE,
      declaredProjectDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/terraform/best-practice`,
    });
    expect(project.checks.length).toEqual(7);
    expect(project).toMatchSnapshot();
  });
  it('should define correctly for a bad practice', async () => {
    const project = await getProjectCheckDeclaration({
      purpose: FileCheckPurpose.BAD_PRACTICE,
      declaredProjectDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times/bad-practices/moment`,
    });
    expect(project.checks.length).toEqual(1);
    expect(project).toMatchSnapshot();
  });
  it('should still have the fix functions defined on the file checks', async () => {
    const project = await getProjectCheckDeclaration({
      purpose: FileCheckPurpose.BEST_PRACTICE,
      declaredProjectDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/prettier/best-practice`,
    });
    expect(project.checks[0].fix).toBeDefined();
    expect(project.checks[0].fix).not.toBeNull();
    expect(project).toMatchSnapshot();
  });
});
