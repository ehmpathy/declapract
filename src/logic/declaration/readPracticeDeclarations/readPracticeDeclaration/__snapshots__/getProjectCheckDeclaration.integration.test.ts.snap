// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProjectCheckDeclarationFromDirectory should define correctly for a bad practice 1`] = `
ProjectCheckDeclaration {
  "checks": Array [
    FileCheckDeclaration {
      "check": [Function],
      "fix": null,
      "pathGlob": "package.json",
      "purpose": "BAD_PRACTICE",
      "required": true,
      "type": "CUSTOM",
    },
  ],
  "name": "moment",
  "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
}
`;

exports[`getProjectCheckDeclarationFromDirectory should define correctly when many files in many levels 1`] = `
ProjectCheckDeclaration {
  "checks": Array [
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": ".gitignore",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "CONTAINS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": ".terraform-version",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "EQUALS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": "provision/terraform/product/variables.tf",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "EQUALS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": "provision/terraform/environments/dev/main.tf",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "CONTAINS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": "provision/terraform/environments/dev/versions.tf",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "EQUALS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": "provision/terraform/environments/prod/main.tf",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "CONTAINS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": "provision/terraform/environments/prod/versions.tf",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "EQUALS",
    },
  ],
  "name": "terraform",
  "readme": "best pattern found for terraform so far is to put different environments into their own directory - and just have them reference a shared product dir
",
}
`;

exports[`getProjectCheckDeclarationFromDirectory should define correctly when only one root level file check and a readme 1`] = `
ProjectCheckDeclaration {
  "checks": Array [
    FileCheckDeclaration {
      "check": [Function],
      "fix": null,
      "pathGlob": "package.json",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "CUSTOM",
    },
  ],
  "name": "dates-and-times",
  "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
}
`;

exports[`getProjectCheckDeclarationFromDirectory should still have the fix functions defined on the file checks 1`] = `
ProjectCheckDeclaration {
  "checks": Array [
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": ".prettierignore",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "EQUALS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": "package.json",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "CONTAINS",
    },
    FileCheckDeclaration {
      "check": [Function],
      "fix": [Function],
      "pathGlob": "prettier.config.js",
      "purpose": "BEST_PRACTICE",
      "required": true,
      "type": "EQUALS",
    },
  ],
  "name": "prettier",
  "readme": null,
}
`;
